### PR TITLE
Introduce exception for concurrent migration and test on map destroy in OnHeapMapScanP

### PR DESCRIPTION
**Pull request by Fly-Style**
Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Request reviewers if possible

### CODE REVIEW

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java [hunk 1]
```diff
@@ -262,6 +290,170 @@ public void test_filterMatchingOneRow() {
...
+                .jobConfig(new JobConfig().setArgument(SQL_ARGUMENTS_KEY_NAME, emptyList()))
+                .outputChecker(LENIENT_SAME_ITEMS_ANY_ORDER)
+                .disableSnapshots()
+                .disableProgressAssertion()
+                .expectOutput(expected);
+
+        // Destroy the map.
+        instance1.getMap(MAP_OBJECT).destroy();
+        try {
+            Thread.sleep(50L);
```
Comment (gierlachg) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java:437]: I would not even try to rely on `sleep()` here. We need better way to ensure `destroy()` is concurrent to map scan.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java [hunk 2]
```diff
@@ -262,6 +290,170 @@ public void test_filterMatchingOneRow() {
...
+            }
+        }
+
+        assertNotEquals(-1, localPartition);
+
+        // Load entries into the partition such that is spans several batches.
+        int loaded = 0;
+        int currentKey = 0;
+
+        while (loaded < BATCH_SIZE * 3 / 2) {
```
[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java:322]: 1024, it's a constant

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java [hunk 3]
```diff
@@ -262,6 +290,170 @@ public void test_filterMatchingOneRow() {
                 .expectOutput(expected);
     }
 
+    @Test
+    public void testConcurrentMigration() throws Exception {
+        final TestHazelcastInstanceFactory FACTORY = new TestHazelcastInstanceFactory(2);
```
[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java:295]: Nope, it was introduce only for manual instance creation.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gierlachg) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java:322]: But why it is 1024? Why it has to be 3/2 * 1024 entries? Is it somehow connected to some runtime setting? What if it changes?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java (hunk 1 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java:437]: Any suggestions? For me, it's the simplest way to wait until map will be destroyed.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java:322]: Ah, I got you. It was done in older way for non-actual reasons atm, so right now batch size doesn't matter.
Thanks for catching this.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java (hunk 1 above)
Comment (gierlachg) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java]: But it does not guarantee that the map will be destroyed before the verification (I understand this is what you are trying to achieve here) + it slows down the test. What about `MapServiceContext.getMapContainers().keySet()`?

**Comment by Fly-Style** [comment]
Since we re-prioritize the tasks and temporarily abandon the existing scan processor with direct map access for short term, I just ignore the test and would like postpone its implementation since it exceeded my time estimation.

**Pull request closed without merging**